ArjunaJTS/jtax/classes/com/arjuna/ats/internal/jta/resources/jts/orbspecific/XAResourceRecord.java

```diff
@@ -422,7 +432,8 @@ public void rollback() throws org.omg.CORBA.SystemException,
 				{
 				    _phaseTwoStarted = true;
 					if (_prepared)
-						destroyState();
+						if (destroyState)
+							destroyState();
```

Comment:
I think I can see a problem with this, by not calling destroyState we are then not going to call removeConnection which is going to keep at least the JDBC transactional driver associated with this XAResourceRecord. I am not sure if that is the cause of the failures that @marcosgopen shared though because I am not sure if it's expected a resource manager in the test (not necessarily the database) would be returning RMFAIL.

Reply:
I think it was a missing call to `removeConnection()` but it wasn't quite for the reason I thought. By adding the code as I added it, the `removeConnection()` had moved to being an else clause on the `if (destroyState)` rather than being the else of the `if (_prepared)`